### ISSUE TITLE
Bug in CosineAnnealingLR (division by zero)

### ISSUE THREAD

**Issue by ptrblck** [pytorch/pytorch#17913]
## 🐛 Bug

The `CosineAnnealingLR` throws a `ZeroDivisionError` after being applied for more than one epoch.
Also, the learning rates returned by the `CosineAnnealingLR` do not match the internal learning rates of the passed `optimizer`.

## To Reproduce
```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim.lr_scheduler import _LRScheduler
import math

# Test new scheduler
model = nn.Linear(10, 2)
optimizer = optim.SGD(model.parameters(), lr=1.)
steps = 10
scheduler = optim.lr_scheduler.CosineAnnealingLR(optimizer, steps)

try:
    for epoch in range(2):
        for idx in range(steps):
            scheduler.step()
            print(scheduler.get_lr())
            print(optimizer.param_groups[0]['lr'])
except ZeroDivisionError as e:
    print(e)

> [1.0]
1.0
[0.9516553824444453]
0.9755282581475768
[0.8386590697412432]
0.9045084971874737
[0.6968044012954536]
0.7938926261462366
[0.5395961100811341]
0.6545084971874737
[0.38196601125010515]
0.5
[0.23872875703131577]
0.34549150281252633
[0.12295598939793909]
0.2061073738537635
[0.04424211972088169]
0.09549150281252633
[0.006271407734229157]
0.024471741852423234
[0.0]
0.0
float division by zero
```

## Expected behavior

Using the legacy implementation, no exception is raised and the learning rates of the scheduler and optimizer are equal:
```python
# Compare with legacy implementation
class LegacyCosineAnnealingLR(_LRScheduler):
    r"""Set the learning rate of each parameter group using a cosine annealing
    schedule, where :math:`\eta_{max}` is set to the initial lr and
    :math:`T_{cur}` is the number of epochs since the last restart in SGDR:

    .. math::

        \eta_t = \eta_{min} + \frac{1}{2}(\eta_{max} - \eta_{min})(1 +
        \cos(\frac{T_{cur}}{T_{max}}\pi))

    When last_epoch=-1, sets initial lr as lr.

    It has been proposed in
    `SGDR: Stochastic Gradient Descent with Warm Restarts`_. Note that this only
    implements the cosine annealing part of SGDR, and not the restarts.

    Args:
        optimizer (Optimizer): Wrapped optimizer.
        T_max (int): Maximum number of iterations.
        eta_min (float): Minimum learning rate. Default: 0.
        last_epoch (int): The index of last epoch. Default: -1.

    .. _SGDR\: Stochastic Gradient Descent with Warm Restarts:
        https://arxiv.org/abs/1608.03983
    """

    def __init__(self, optimizer, T_max, eta_min=0, last_epoch=-1):
        self.T_max = T_max
        self.eta_min = eta_min
        super(LegacyCosineAnnealingLR, self).__init__(optimizer, last_epoch)

    def get_lr(self):
        return [self.eta_min + (base_lr - self.eta_min) *
                (1 + math.cos(math.pi * self.last_epoch / self.T_max)) / 2
                for base_lr in self.base_lrs]


model = nn.Linear(10, 2)
optimizer = optim.SGD(model.parameters(), lr=1.)
steps = 10
scheduler = LegacyCosineAnnealingLR(optimizer, steps)

for epoch in range(2):
    for idx in range(steps):
        scheduler.step()
        print(scheduler.get_lr())
        print(optimizer.param_groups[0]['lr'])

> [1.0]
1.0
[0.9755282581475768]
0.9755282581475768
[0.9045084971874737]
0.9045084971874737
[0.7938926261462366]
0.7938926261462366
[0.6545084971874737]
0.6545084971874737
[0.5]
0.5
[0.34549150281252633]
0.34549150281252633
[0.2061073738537635]
0.2061073738537635
[0.09549150281252633]
0.09549150281252633
[0.024471741852423234]
0.024471741852423234
[0.0]
0.0
[0.024471741852423123]
0.024471741852423123
[0.09549150281252622]
0.09549150281252622
[0.20610737385376338]
0.20610737385376338
[0.3454915028125262]
0.3454915028125262
[0.4999999999999999]
0.4999999999999999
[0.6545084971874736]
0.6545084971874736
[0.7938926261462365]
0.7938926261462365
[0.9045084971874737]
0.9045084971874737
[0.9755282581475768]
0.9755282581475768
```

## Environment

 - PyTorch Version (e.g., 1.0): `1.0.0.dev20190312`
 - OS (e.g., Linux): Ubuntu 18.04
 - How you installed PyTorch (`conda`, `pip`, source): conda nightlies
 - Python version: 3.7

## Additional context

This issue might to be related to the refactoring of all schedulers done [here](https://github.com/pytorch/pytorch/pull/14010).

CC @chandlerzuo @ezyang

**Comment by Kindpire** [comment]
Same problem here, also with 1.0.0.dev20190312 (conda nightly) but python 3.6 in ubuntu 16.04

**Comment by ezyang** [comment]
I will look into this.

**Comment by ezyang** [comment]
I've confirmed that reverting #14010 fixes the problem. I'm going to revert it for now and then investigate more.
